### DISCUSSION TITLE
Usage of nested partial folders

### DISCUSSION THREAD

**Discussion by Chris2011** [alexlafroscia/vite-plugin-handlebars#30] [Ideas]
As already mentioned here: #26 it would be cool, to have nested partial folders. In my case, I have not only a header.hbs file, also an additional css file.

Folder structure could be this (it is currently in my case):

```
/partials
   /common
   /header
      header.hbs
      _header.css
   /footer
   ...
```

I don't know, maybe not only one deep but also 2 or 3 but this is not my case, could be for someone else if the project is bigger. I think it is ok, because this happens at the build time so no worries about performance.

To identify the correct partials, it is needed to map the folder as a namespace before the partial name:

`{{> header/header }}` or `{{> common/foo}}` this is working as you can see it here on line 17 as an example: https://plnkr.co/edit/VJx5uMNiDPeep9kx. So there shouldn't be any trouble with that. Later we can think about a mapping but this is not that necessary I think.

**Comment by alexlafroscia** [comment]
I think that sounds like a great idea!

I totally agree that it would make sense to support an arbitrary amount of nesting and building up whatever kind of complex namespace for the partial that the user wants. That shouldn't be that much harder than going one level deep; just need a little bit of recursion!

**Reply by Chris2011** [reply]
Yeah, think so too, your code is really yimple, yet :D

**Comment by alexlafroscia** [comment]
I implemented this here https://github.com/alexlafroscia/vite-plugin-handlebars/commit/9e657b2fecc045eb0d1b49a3304cde585522aa05

It was released as part of `1.2.0` just now!

**Comment by Chris2011** [comment]
Super cool, thx :)

**Comment by Chris2011** [comment]
Unfortunately it doesn't work with 1.2.0 and 1.3.0. Maybe I missed smth, this is my config from the readme:

```
import { resolve } from 'path';
import handlebars from 'vite-plugin-handlebars';

export default {
  plugins: [
    handlebars({
      partialDirectory: resolve(__dirname, 'partials'),
    }),
  ],
};
```

`{{>head/head}}``this is in my startpage.html.

When I do a `yarn build` I got this error:

```
[vite:build-html] The partial head could not be found
file: C:/Projekte/Others/smth/pages/startpage/startpage.html
error during build:
Error: The partial head/head could not be found
```

**Reply by alexlafroscia** [reply]
What does your `partials` directory look like? What's the file layout?

**Reply by alexlafroscia** [reply]
[This test](https://github.com/alexlafroscia/vite-plugin-handlebars/blob/9a095e3d8fbdf76c1c99a324bb17626577f1cb2f/__tests__/partials.ts#L58-L73) was written against sub-directories of the `partials` directory, and suggests that things _should_ be working...

**Reply by Chris2011** [reply]
Here is how my partials directory looks lik:

![image](https://user-images.githubusercontent.com/795658/111991152-45b48f80-8b14-11eb-8a7f-128a3d8313dc.png)

The file layout of my startpage.html looks like this:

```html
{{>head/head}} {{>header/header}}

<main id="start">
  <section id="banner"></section>
  <section>
    <div class="inner padding-standard">
...
```

Let me know, if you need more information :)

**Reply by Chris2011** [reply]
And this is how my main.js file looks like:

```javascript
// Global CSS.
import './style.css';

import './pages/startpage/startpage.html';
```

**Reply by alexlafroscia** [reply]
Can you share a minimal reproduction repo for me? I would love to solve this, but I'm not sure there's much I can do without the code...

**Reply by alexlafroscia** [reply]
If the project you're working on is public, or you feel comfortable inviting me to the private repo temporarily, I'm happy to pull your app and dig into it

**Reply by Chris2011** [reply]
Sure, will do a sample project later :)

**Reply by alexlafroscia** [reply]
Thanks!

Let's continue this conversation on #41 -- I made an issue from this conversation since we're investigating whether there's a bug here

**Comment by McGern** [comment]
Hi, just chipping in much later down the track.

I'm using the library with nested partials and all is working, and was looking back at the original ticket, then this discussion, and agree with the reasoning of having the nested partial path reflected in the partial name. But it does become quite verbose especially when you are using partial blocks e.g.

Instance where there is a components folder with multiple sub folders all containing named components within a same named folder

{{#> core-components/wrapper/wrapper }}
     inside
{{/core-components/wrapper/wrapper}}

For the (subjective) majority, you wouldn't be naming partials with the same name in a different folders because that just make life harder during development, so would it be possible to add an option so it doesn't nest the partial names (I think this was mentioned in the original ticket)? 

If there is a conflict, suggest that it just ignores the namespace of any subsequent partials found, and that becomes something for the developer to manage i.e. you can't have both partials with the same name, and shorted mappings. The current functionality already does something similar, so if you have button.html and button.hbs in the same folder, it will favour the html version.
EDIT: Looks like it just writes over the top of existing partial when registerPartial happens for a name collision.

Maybe even have both mappings available? So you can access by the full path, and by the partial name, using the logic above to ignore the duplicated instances when shortening.

Am interested in anyone's thoughts.

Thanks,
J
